In [98]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib
import matplotlib.pyplot as plt
import warnings

In [99]:
#Load data for train and test
train = pd.read_csv('C:/Users/ozhat/Desktop/hw1_fold/data.csv')

In [100]:
train.head()

Id  LotFrontage  LotArea  YearBuilt  YearRemodAdd  TotalBsmtSF  1stFlrSF  \
0   1         65.0     8450       2003          2003          856       856   
1   2         80.0     9600       1976          1976         1262      1262   
2   3         68.0    11250       2001          2002          920       920   
3   4         60.0     9550       1915          1970          756       961   
4   5         84.0    14260       2000          2000         1145      1145   

   2ndFlrSF  GrLivArea  FullBath  ...  KitchenAbvGr  TotRmsAbvGrd  Fireplaces  \
0       854       1710         2  ...             1             8           0   
1         0       1262         2  ...             1             6           1   
2       866       1786         2  ...             1             6           1   
3       756       1717         1  ...             1             7           1   
4      1053       2198         2  ...             1             9           1   

   GarageCars  GarageArea  OpenPorchSF  EnclosedPorch  PoolArea  YrSold  \
0           2         548           61              0         0    2008   
1           2         460            0              0         0    2007   
2           2         608           42              0         0    2008   
3           3         642           35            272         0    2006   
4           3         836           84              0         0    2008   

   SalePrice  
0     208500  
1     181500  
2     223500  
3     140000  
4     250000  

[5 rows x 22 columns]

In [101]:
train.shape

(1460, 22)

In [102]:
n = train.shape[0] # number of samples
d = train.shape[1] # number of columns/attrinutes
# the last column is y, the lables (the real prices)

## let's check the type of each columns 
## (there are NaNs as well)

In [103]:
#you can also check the data set information using the info() command. 
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 22 columns):
Id               1460 non-null int64
LotFrontage      1201 non-null float64
LotArea          1460 non-null int64
YearBuilt        1460 non-null int64
YearRemodAdd     1460 non-null int64
TotalBsmtSF      1460 non-null int64
1stFlrSF         1460 non-null int64
2ndFlrSF         1460 non-null int64
GrLivArea        1460 non-null int64
FullBath         1460 non-null int64
HalfBath         1460 non-null int64
BedroomAbvGr     1460 non-null int64
KitchenAbvGr     1460 non-null int64
TotRmsAbvGrd     1460 non-null int64
Fireplaces       1460 non-null int64
GarageCars       1460 non-null int64
GarageArea       1460 non-null int64
OpenPorchSF      1460 non-null int64
EnclosedPorch    1460 non-null int64
PoolArea         1460 non-null int64
YrSold           1460 non-null int64
SalePrice        1460 non-null int64
dtypes: float64(1), int64(21)
memory usage: 251.1 KB


# Q1. 

## Create from the dataframe train a design Matrix X and a vector y 

## you can do in in few steps:

In [104]:
df=train.copy()

In [105]:
df=df.dropna()

In [106]:
y =df['SalePrice']

In [107]:
df=df.drop(['Id', 'SalePrice'], axis=1) 

In [108]:
average_prices=y.mean()

In [109]:
y=y.values.reshape(-1,1)

In [110]:
binary_col=[]
for i in range(len(y)):
    test=y[i]/average_prices
    if (test>=0.8) and (test<=1.2):
        binary_col.append(1)
    else:
        binary_col.append(0)
df['binary_col'] = binary_col 

In [111]:
df.head()

LotFrontage  LotArea  YearBuilt  YearRemodAdd  TotalBsmtSF  1stFlrSF  \
0         65.0     8450       2003          2003          856       856   
1         80.0     9600       1976          1976         1262      1262   
2         68.0    11250       2001          2002          920       920   
3         60.0     9550       1915          1970          756       961   
4         84.0    14260       2000          2000         1145      1145   

   2ndFlrSF  GrLivArea  FullBath  HalfBath  ...  KitchenAbvGr  TotRmsAbvGrd  \
0       854       1710         2         1  ...             1             8   
1         0       1262         2         0  ...             1             6   
2       866       1786         2         1  ...             1             6   
3       756       1717         1         0  ...             1             7   
4      1053       2198         2         1  ...             1             9   

   Fireplaces  GarageCars  GarageArea  OpenPorchSF  EnclosedPorch  PoolArea  \
0           0           2         548           61              0         0   
1           1           2         460            0              0         0   
2           1           2         608           42              0         0   
3           1           3         642           35            272         0   
4           1           3         836           84              0         0   

   YrSold  binary_col  
0    2008           1  
1    2007           1  
2    2008           0  
3    2006           0  
4    2008           0  

[5 rows x 21 columns]

In [112]:
new_y=df['binary_col'].copy()
new_y=new_y.values.reshape(-1,1)
df=df.drop(['binary_col'], axis=1) 

In [113]:
X = df.to_numpy() 

In [114]:
y = new_y

# Q2. 

## Create from X and the vector y:

### 1. Estimations of the weights w_hat

### 2. Estimations of y_hat

### 3. mean square error?

### 4. mean absolute error?

### 5. Calculate the matrix that takes w to x_hat

In [115]:
from sklearn.linear_model import LogisticRegression
logmodel1 = LogisticRegression()
logmodel1.fit(X,y)
w_hat=logmodel1.coef_[0]
w_hat.shape

C:\Users\ozhat\Miniconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\ozhat\Miniconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


(20,)

In [116]:
y_hat=logmodel1.predict(X)
y_hat

array([0, 1, 1, ..., 0, 0, 0], dtype=int64)

In [117]:
mean_square_error = np.square(np.subtract(y,y_hat)).mean() 
mean_square_error

0.42212117157434026

In [118]:
mean_average_error = np.abs((y_hat-y)).mean()
mean_average_error

0.42212117157434026

In [119]:
w_hat

array([ 2.26884790e-03, -1.04175093e-05, -1.02058097e-05,  1.39471583e-02,
       -1.80996759e-04, -2.27617454e-03, -3.11991339e-03,  1.10274906e-03,
        1.17959598e+00,  1.38584798e+00,  4.50020746e-01, -5.91986495e-02,
       -1.52359996e-01,  3.47136321e-01,  6.08980583e-01, -1.83125062e-03,
       -3.17386920e-04,  1.81521165e-04,  2.90635274e-03, -1.45306631e-02])

In [120]:
# Q3. 

## Using the weights you got :

### 1. Estimations of the 3 most influential features?

### 2. Estimations of the 3 less influential features?


In [121]:
from heapq import nlargest,nsmallest
val_large= (nlargest(3, np.abs(w_hat)))
influential_features=[]
index = 0
while index < len(val_large):
   influential_features.append(df.columns[list(np.abs(w_hat)).index(val_large[index])])
   index+=1
val_small= (nsmallest(3, np.abs(w_hat)))
neglectible_features=[]
index = 0
while index < len(val_small):
   neglectible_features.append(df.columns[list(np.abs(w_hat)).index(val_small[index])])
   index+=1


In [122]:
influential_features

['HalfBath', 'FullBath', 'GarageCars']

In [123]:
neglectible_features

['YearBuilt', 'LotArea', 'TotalBsmtSF']

# Q4. 

## Take the same regression problem but now with only 3 features that you found in previous question:

### 1. w_hat?

### 2. y_hat?

### 3. mean sum square error?

### 4. mean sum absolute error?



In [124]:
X3 =df[influential_features]

In [125]:
# calc new design matrix, don't overide the onld matrix X
X3=X3.to_numpy()

In [126]:
logmodel2 = LogisticRegression()
logmodel2.fit(X3,y)
w3_hat=logmodel2.coef_[0]
w3_hat.shape

C:\Users\ozhat\Miniconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\ozhat\Miniconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


(3,)

In [127]:
w3_hat

array([0.69000475, 0.54439305, 0.04737262])

In [128]:
y3_hat=logmodel2.predict(X3)

In [129]:
y3_hat

array([1, 0, 1, ..., 0, 0, 0], dtype=int64)

In [130]:
mean_square_error3 = np.square(np.subtract(y,y3_hat)).mean()

In [131]:
mean_square_error3

0.42378575722007955

In [132]:
mean_average_error3 = np.abs((y3_hat-y)).mean()

In [133]:
mean_average_error3

0.42378575722007955

# Q5. 

## Repeat Q2 but now with only half of the samples 

### (sampled randomly for the rows of the original X and y which had 1460 rows)

### 1. Estimations of the weights w_hat

### 2. Estimations of y_hat

### 3. mean square error?

### 4. mean absolute error?

### 5. Calculate the matrix that takes w to x_hat

In [134]:
df=train[train['Id']<=n/2]
df=df.dropna()
df.shape

(606, 22)

In [135]:
df

Id  LotFrontage  LotArea  YearBuilt  YearRemodAdd  TotalBsmtSF  \
0      1         65.0     8450       2003          2003          856   
1      2         80.0     9600       1976          1976         1262   
2      3         68.0    11250       2001          2002          920   
3      4         60.0     9550       1915          1970          756   
4      5         84.0    14260       2000          2000         1145   
..   ...          ...      ...        ...           ...          ...   
724  725         86.0    13286       2007          2008         1698   
725  726         60.0     6960       1970          1970          864   
727  728         64.0     7314       2007          2007         1232   
728  729         85.0    11475       1958          1958         1584   
729  730         52.0     6240       1925          1950          780   

     1stFlrSF  2ndFlrSF  GrLivArea  FullBath  ...  KitchenAbvGr  TotRmsAbvGrd  \
0         856       854       1710         2  ...             1             8   
1        1262         0       1262         2  ...             1             6   
2         920       866       1786         2  ...             1             6   
3         961       756       1717         1  ...             1             7   
4        1145      1053       2198         2  ...             1             9   
..        ...       ...        ...       ...  ...           ...           ...   
724      1698         0       1698         2  ...             1             8   
725       864         0        864         1  ...             1             5   
727      1232         0       1232         2  ...             1             6   
728      1776         0       1776         2  ...             2             9   
729       848         0       1208         1  ...             1             5   

     Fireplaces  GarageCars  GarageArea  OpenPorchSF  EnclosedPorch  PoolArea  \
0             0           2         548           61              0         0   
1             1           2         460            0              0         0   
2             1           2         608           42              0         0   
3             1           3         642           35            272         0   
4             1           3         836           84              0         0   
..          ...         ...         ...          ...            ...       ...   
724           1           3         768           64              0         0   
725           0           2         660            0              0         0   
727           0           2         632            0              0         0   
728           0           3         888           25              0         0   
729           0           2         539           23            112         0   

     YrSold  SalePrice  
0      2008     208500  
1      2007     181500  
2      2008     223500  
3      2006     140000  
4      2008     250000  
..      ...        ...  
724    2009     320000  
725    2009     120500  
727    2009     194500  
728    2009     110000  
729    2009     103000  

[606 rows x 22 columns]

In [136]:
y5 =df['SalePrice']

In [137]:
df=df.drop(['Id', 'SalePrice'], axis=1)



In [138]:
# X5=X5.to_numpy()

In [139]:
y5=y5.to_numpy()

In [140]:
binary_col=[]
average_prices=y5.mean()
for i in range(len(y5)):
    test=y5[i]/average_prices
    if (test>=0.8) and (test<=1.2):
        binary_col.append(1)
    else:
        binary_col.append(0)
df['binary_col'] = binary_col 

In [141]:
y5=df['binary_col'].copy()
y5=y5.values.reshape(-1,1)
df=df.drop(['binary_col'], axis=1) 

In [142]:
X5=df.to_numpy()

In [143]:
logmodel3 = LogisticRegression()
logmodel3.fit(X5,y5)
w5_hat=logmodel3.coef_[0]
w5_hat.shape

C:\Users\ozhat\Miniconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\ozhat\Miniconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


(20,)

In [144]:
y5_hat=logmodel3.predict(X5)
y5_hat

array([0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0,
       1, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       0, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1,
       1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1,
       0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0,
       0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0,
       0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0,

In [145]:
mean_square_error5 = np.square(np.subtract(y5,y5_hat)).mean()

In [146]:
mean_square_error5

0.3905989608861876

In [147]:
mean_average_error5 =np.abs((y5_hat-y5)).mean()

In [148]:
mean_average_error5

0.3905989608861876

In [149]:
# Q6. 

## Compare the results between all 3 models that you built

In [150]:
models_list=['m1-original','m2-3 columns','m3-half samples']

In [151]:
arr_mse=[mean_square_error,mean_square_error3,mean_square_error5]

In [152]:
index_sorted_mse=np.argsort(arr_mse)
index_smallest_mse=index_sorted_mse[0]
index_medium_mse=index_sorted_mse[1]
index_biggest_mse=index_sorted_mse[2]
print("The smallest MSE:",arr_mse[index_smallest_mse]," of ", models_list[index_smallest_mse] )
print("The medium MSE:",arr_mse[index_medium_mse]," of ", models_list[index_medium_mse] )
print("The biggest MSE:",arr_mse[index_biggest_mse]," of ", models_list[index_biggest_mse] )

The smallest MSE: 0.3905989608861876  of  m3-half samples
The medium MSE: 0.42212117157434026  of  m1-original
The biggest MSE: 0.42378575722007955  of  m2-3 columns
